internal/xfn/cached/cached_runner.go

```diff
@@ -0,0 +1,383 @@
+/*
+Copyright 2025 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License. You may obtain a copy of the
+License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+// Package cached implements a cached function runner.
+package cached
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+	"io/fs"
+	"path/filepath"
+	"time"
+
+	"github.com/spf13/afero"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	fnv1 "github.com/crossplane/crossplane/apis/apiextensions/fn/proto/v1"
+)
+
+// A CacheMissReason indicates what caused a cache miss.
+type CacheMissReason string
+
+// Cache miss reasons.
+const (
+	ReasonNotCached       CacheMissReason = "NotCached"
+	ReasonDeadlineExpired CacheMissReason = "DeadlineExpired"
+	ReasonEmptyRequestTag CacheMissReason = "EmptyRequestTag"
+	ReasonEmptyResponse   CacheMissReason = "EmptyResponse"
+	ReasonError           CacheMissReason = "Error"
+)
```

Comment:
In retrospect I'm not sure these are very useful. Maybe for debug logs, but not for metrics?

In particular `DeadlineExpired` could be misleading. A cache miss due to an expired deadline could just as easily have reason `NotFound` if the garbage collector deleted it before Crossplane sent the request.

Reply:
Pushed a commit that drops these from metrics.